interfaces/udev/backend.go

```diff
@@ -0,0 +1,99 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2016 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+// Package udev implements integration between snappy, udev and
+// ubuntu-core-laucher around tagging character and block devices so that they
+// can be accessed by applications.
+//
+// TODO: Document this better
+package udev
+
+import (
+	"bytes"
+	"fmt"
+
+	"github.com/ubuntu-core/snappy/dirs"
+	"github.com/ubuntu-core/snappy/interfaces"
+	"github.com/ubuntu-core/snappy/osutil"
+	"github.com/ubuntu-core/snappy/snap"
+)
+
+// Backend is responsible for maintaining udev rules.
+type Backend struct{}
+
+// Configure creates udev rules specific to a given snap.
+// If any of the rules are changed or removed then udev database is reloaded.
+//
+// Since udev has no concept of a complain mode, developerMode is ignored.
+func (b *Backend) Configure(snapInfo *snap.Info, developerMode bool, repo *interfaces.Repository) error {
+	snippets, err := repo.SecuritySnippetsForSnap(snapInfo.Name, interfaces.SecurityUDev)
+	if err != nil {
+		return fmt.Errorf("cannot obtain udev security snippets for snap %q: %s", snapInfo.Name, err)
+	}
+	content, err := b.combineSnippets(snapInfo, snippets)
+	if err != nil {
+		return fmt.Errorf("cannot obtain expected udev rules for snap %q: %s", snapInfo.Name, err)
+	}
+	glob := fmt.Sprintf("70-%s.rules", interfaces.SecurityTagGlob(snapInfo))
+	changed, removed, err := osutil.EnsureDirState(dirs.SnapUdevRulesDir, glob, content)
+	if err != nil {
+		return fmt.Errorf("cannot synchronize udev rules for snap %q: %s", snapInfo.Name, err)
+	}
+	if len(changed) > 0 || len(removed) > 0 {
+		return ReloadRules()
+	}
+	return nil
+}
+
+// Deconfigure removes udev rules specific to a given snap.
+// If any of the rules are removed then udev database is reloaded.
+//
+// This method should be called after removing a snap.
+func (b *Backend) Deconfigure(snapInfo *snap.Info) error {
+	glob := fmt.Sprintf("70-%s.rules", interfaces.SecurityTagGlob(snapInfo))
+	changed, removed, err := osutil.EnsureDirState(dirs.SnapUdevRulesDir, glob, nil)
+	if err != nil {
+		return fmt.Errorf("cannot synchronize udev rules for snap %q: %s", snapInfo.Name, err)
```

Comment:
Per prior note, what's the value of the partial result if we're ignoring it? Also, what's the plan for fixing the bad data?